### DISCUSSION TITLE
メモアプリの運用について

### DISCUSSION THREAD

**Discussion by kuri0616** [Progaku-copy/progaku-archive#84] [General]
メモアプリの運用方法を確定し、仕様を固める

## テーマ
リリース後、アーカイブに残したいものを初めは、手動でメモアプリの方に保存していくことになる。この追加作業を誰がどのように行うかによって仕様が変わるため確認をしたい。
(リリース後、SlackAPIから取得し反映できる機能は実装したいと考えている)

## 案
僕からは２つ案を考えたのでご意見いただきたいです！

### 案1「管理者権限を持つ、ユーザがメモアプリへの追加作業を行うようにする」
管理者用ユーザ、閲覧用のユーザのみ作成する。
閲覧したい人は、閲覧用の共用のユーザID、パスワードを使用して見る。
#### メリット
* 閲覧用ユーザ、管理者用ユーザだけになるのでユーザの管理は楽になる(一人ひとり付与すると結構大変な気がした)

#### デメリット
* アーカイブ追加作業など、管理の担当者の手間がかかる

#### 案1にした場合の仕様変更
現状、ログインしているユーザの情報(名前など）でメモを作成するような仕様になっているため、誰がSlack上で投稿したものなのかはわからなくなる。もしわかるようにするならメモテーブルに「作成者」等、Slack上で誰が投稿したかの情報を格納するカラムを追加する必要があると考えられる。

### 案2「コミュニティのユーザ一人ひとりにアカウントを付与し、誰もが作成できるようにする」(当初の予定)
一人ひとりのユーザ名、パスワードを作成する。
自己紹介アプリと同じ運用？？

#### メリット
* 誰もが投稿できるので、この情報は残したいと思った時、手軽に追加ができる。

#### デメリット
* ユーザ情報の管理は複雑になりそう

#### 案2にした場合の仕様変更
* Slackで作成した本人が、アーカイブに追加する運用なら変更無
* 誰でも作成できるようにするなら案１同様に「作成者」等、追加する必要がありそう

僕の個人的な意見としては、案1が良いのかなと思っていたりします！
(えーきちさんと打ち合わせした時と考え変わっていてごめんなさい笑)
理由としては、閲覧するまでの手軽さが大切と感じたこと、アーカイブに残す頻度を考慮すると管理者側の負担もそこまで大きくないと思ったからです！
見たいとなったとき、たまにしか入力しないユーザ名、パスワードだと忘れてしまったり、手間が増えることで余り活用されなくなったりすると勿体無いし、Slackの更新頻度見ても、アーカイブに移行する作業ってそこまで多くないと思ったので。

**Comment by yuto-ng** [comment]
検討ありがとうございます！
改めて考えたのですが、

> Slackの更新頻度見ても、アーカイブに移行する作業ってそこまで多くない

これは確かにそこまで多くなさそうなので、追加の手軽さよりも閲覧の手軽さを優先でいいと思いました！
ただ、閲覧側の気持ちとしては情報だけでなく誰の発信なのかって結構大事になるのかなとも思うので、作成時に誰が発信した情報なのかはわかるようにしたいです！(この場合「作成者」って表現ではなくなる...？)

**Reply by kuri0616** [reply]
> ただ、閲覧側の気持ちとしては情報だけでなく誰の発信なのかって結構大事になるのかなとも思うので、作成時に誰が発信した情報なのかはわかるようにしたいです！(この場合「作成者」って表現ではなくなる...？)

そうですね！これに関しては、とくゆーさんが

> 単純にメモ作成時に名前の項目を追加して入力するだけにすればシンプルかなっと。

と記載されているように「名前」の項目追加で入力するようにすれば実現できそうですね！

**Comment by yuta-tokuda** [comment]
案２のケースでログイン機能なしというのは何かデメリットありそうですか？？

> たまにしか入力しないユーザ名、パスワードだと忘れてしまったり、手間が増えることで余り活用されなくなったり

これは自分も思うので
単純にメモ作成時に名前の項目を追加して入力するだけにすればシンプルかなっと。

もしユーザー別で検索等したいのであればUserとの紐付けが必要になるとは思いますが・・・

**Reply by kuri0616** [reply]
> 案２のケースでログイン機能なしというのは何かデメリットありそうですか？？

デメリットでいうと、おっちーさんの言うように、Progaku以外の人にアクセスされた場合にその人たちに、自由に作成、編集されてしまうという恐れがあるのかなと思いました！

> 単純にメモ作成時に名前の項目を追加して入力するだけにすればシンプルかなっと。

確かに、これが一番シンプルで良さそうです。今後Slackから取得等実装した際も、挿入しやすいですね！

> もしユーザー別で検索等したいのであればUserとの紐付けが必要になるとは思いますが・・・

これに関しては、アーカイブが増えていくにつれて、必要性感じた際に、追加とかでも良さそうですね！

**Reply by yuta-tokuda** [reply]
@rikuya98 
>デメリットでいうと、おっちーさんの言うように、Progaku以外の人にアクセスされた場合にその人たちに、自由に作成、編集されてしまうという恐れがあるのかなと思いました！

なるほどです〜

コミュニティメンバーのみの利用なので
>共有のアカウントを使用するようにする(ユーザID:Progaku1234 パスワード:Progaku1234)とか固定のものを作成。

で十分かなとは思いました👀
昔コミュニティ内限定のzoomアカウント等もこんな感じで共通アカウントで使えるようにしてましたし！

**Comment by kakeru-one** [comment]
> 見たいとなったとき、たまにしか入力しないユーザ名、パスワードだと忘れてしまったり、手間が増えることで余り活用されなくなったりすると勿体無いし、Slackの更新頻度見ても、アーカイブに移行する作業ってそこまで多くないと思ったので。

まず、閲覧までの手間を省きたいのであれば、「コミュニティのユーザ一人ひとりにアカウントを付与する方法」でも可能だと思います。
これは、TwitterやInstagramのように、`ログインしないと編集や新規投稿はできないが、閲覧はできる`というような方式を取っても実現できると思ったからです。
（Progaku以外の人もアクセスできてしまっても問題ないか？など考えることはある。もし一部の情報は閲覧できないようにしたいのであれば、公開範囲の`private`と`public`という概念を考えた方が良さそう。）

これを考慮した上で案1, 案2を比較すると、このアプリでSlackのメッセージ履歴以外も気軽に残せるようにしたいのであれば、
案2がいいと思います。
また、新規投稿・編集をしたい場合でに、「たまにしか入力しないユーザ名、パスワードだと忘れてしまったり」の部分はGoogle SSOログインできるようにすると解決できると思います。

次に、このアプリでSlackのメッセージ履歴しか残さないのであれば、どちらでもあまり大差ないので、どちらでもいいと思いました！
（IMOですが、強いていうなら、この場合においてもgoogleアカウントであれば、「email/passwordを忘れる・入力が面倒」といったことはなさそうなので、案2にした上でGoogle SSOログインできるようにするといいかなと思いました。）

**Reply by kuri0616** [reply]
たくさんご意見ありがとうございます！

>  ログインしないと編集や新規投稿はできないが、閲覧はできる
> Progaku以外の人もアクセスできてしまっても問題ないか？など考えることはある

そうなんですよね、これがあったので、閲覧に関しても認証機能が必要なのかなと思ってました💦

> 「たまにしか入力しないユーザ名、パスワードだと忘れてしまったり」の部分はGoogle SSOログインできるようにすると解決できると思います。

確かに、GoogleSSOならこの辺りの手間は大幅に削減できそうですね！
この場合、コミュニティメンバーかどうか判断して制限かけることとかってどうしたらいいんですかね？(Googleアカウントを持っている人なら、誰でも、ログインできるみたいにならないように)

**Reply by kakeru-one** [reply]
@rikuya98 
> この場合、コミュニティメンバーかどうか判断して制限かけることとかってどうしたらいいんですかね？(Googleアカウントを持っている人なら、誰でも、ログインできるみたいにならないように)

アカウントグループ的な概念を導入して、特定のアカウントグループに所属している場合には閲覧できるようにする、
という風にすればいいかなと思いました！
ただ、画面から紐づけるのではなく、internalなAPIもしくはDB作業で紐づけることになりそうですが。
（アカウントグループをログインアカウントの情報を保存しているテーブルに紐づけるイメージ）

**Comment by kuri0616** [comment]
皆さん、コメントありがとうございます！

今回あがってきた点として
* Slackで誰が記載したものなのかわかるような項目の作成
* Slackのメッセージ履歴以外も残すかどうか
* 完全にログイン機能を無くした場合のデメリット
* コミュニティメンバー以外の閲覧を許可するかどうか
がありますが、皆さんどうですかね？

僕としては
### Slackで誰が記載したものなのかわかるような項目の作成
これに関しては、メモテーブルに「名前」のカラムを追加して、作成時に入力するようにできれば良いなと思いました！

### Slackのメッセージ履歴以外も残すかどうか
これに関しては、残したい前提でメモアプリを使うケースはないのかなと思ったので、Slackのメッセージ履歴のみと限定して考えても良いのかなと思いました！(基本、Slackがメインで、残すか、残さないかは第３者から後から判断されるものだと思ったからです)

### 完全にログイン機能を無くした場合のデメリット
これは、コミュニティメンバー以外にアクセスされた際に、自由に作成、編集等ができてしまうことかなと思ったので必要性はあるかなとは思いました！

### コミュニティメンバー以外の閲覧を許可するかどうか
SNS怖いので、コミュニティメンバー内のみが良いのかなと・・・笑

皆さんの意見踏まえて、改めて考えた案としては
* 共有のアカウントを使用するようにする(ユーザID:Progaku1234 パスワード:Progaku1234)とか固定のものを作成。
* メモアプリが完成したら「Slackアーカイブアプリ」みたいなチャンネル作って、使用ルールと共に、ユーザIDとパスワードを記載して、それでログインするように促す
* 上記のユーザIDとパスワードを使用して、誰でもメモの作成、編集などできるようにする
* Slackで誰が投稿したものなのかは、名前の項目に入力して作成するようにする
はどうかなと思っております

固定アカウントで常時周知することでコミュニティメンバーしか閲覧、作成できないようにしつつ、アカウント情報忘れ、手間を最低限に押さえられるかなと
ただ、おっちーさん言うように、GoogleOSS使用した方が手軽ですかね？
